### ISSUE TITLE
Undo not functioning properly.

### ISSUE THREAD

**Issue by LadySaga** [TEdit/Terraria-Map-Editor#1343]
(Sorry for the many posts at once, I just learned where to post bugs, and I really appreciate the work you do, and since I can't code myself, I wanted to help some way at least, by mentioning stuff I've found and couldn't see mentioned when searching the opened topics) :)

So Using Undo and/or Ctrl+Z seems to have 2 alternative results sometimes instead of the (also, and fortunately more often occurring) intended result.

* The most common result is the intended. It undoes the previous action, no further explanation needed.

* Sometimes however if I'm undoing for instance, sprites that I have placed, maybe because I placed them wrong, or maybe a timer turned into a target dummy (see the other post about that), or sometimes I just changed my mind, etc etc etc.
But, when I use Undo (via the menu or by Ctrl+Z makes no difference) it just... doesn't undo it.
If the tiles/walls around it has alternative sprites (such as dungeon bricks and so on) then I still see them update around the sprite, but the sprite is still there.
If I then Redo the action, the sprite disappears.
So in order to undo my sprites, which sometimes work, and when it doesn't, I redo it back, THEN it goes away.
This also messes up the action/undo history if you have previous stuff you wanted to undo as well. Let's say I add 5 lanterns.. I didn't like it so I undo them, but I accidentaly only undo 4 times, and then promptly redo 4 times to remove them, well, that 5th can no longer be undoed, even with the undo/redo trick either. So undo is unreliable with sprites. It works sometimes, but borks some other times.

* The other thing seems more common with tiles and walls, but can happen with sprites as well, just not as common.
Basically, it does Undo... but only partially. And trying to Redo it back doesn't work, and undo more will still keep the thing that failed to undo before, so once this happens, you're pretty much stuck in a point-of-no-return and have to redo/fix the errors manually instead of relying on undo/redo.

* Oh, and almost forgot. In some cases, Undo just doesn't do a single thing. (This is mentioned in the Timer issue as well).

**Comment by Darkstar0001** [comment]
Yeaaaaah... this bug has existed for as long as I can remember. I made a thread about it a long time ago, myself. Hopefully it will really be fixed this time...

**Comment by RussDev7** [comment]
This appears to finally have been fixed in the latest snapshot. I will re-open this if needed.
